### PR TITLE
Make examples searchable by screenshots.

### PR DESCRIPTION
**Pull request by munrocket**
Trying to solve problem with searching page, since we don't have any progress in #18714

[Live page on GitHack](https://rawcdn.githack.com/munrocket/three.js/98454737436eda21bd96a8a2a277d93047f28062/examples/index.html)

2do list:
- [x] ui button that works on mobile phones ( in own PR #18841 )
- [x] spiral sphere geometry that easily to extend with new examples
- [x] MVP with octahedral mapping
- [x] uv maker with stripe mapping
- [x] automatic spiral resize that keep 800x600 ratio
- [x] analytic raycasting
- [x] image picker that works on phones
- [x] fix redirect & uv maker
- [x] highlight tiles
- [x] uv maker with cpu
- [x] highlight names
- [x] preload,  progressive loading

### CODE REVIEW

**Comment by looeee** [comment]
This is very cool ^_^ 

On the other hand, there's a lot to be said for keeping UIs simple. 
The grid of thumbnails idea would certainly be more usable than this, no matter how much you polish it.

However, I'll reserve judgement until you have raycasting implemented. 

By the way, what's analytic raycasting? My searches are not turning up anything much for that phrase.

File: examples/index_searchball.html [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+				var p = new THREE.Vector3();
+				var s = ( - b - Math.sqrt( D ) ) / ( 2 * a );
+				ray.at( s, p );
+				p = new THREE.Spherical().setFromVector3( p );
+
+				p.phi = Math.PI / 2 - p.phi;
+				var t = p.theta / 2 / turns + Math.PI / turns *
+						Math.floor( ( turns * ( Math.PI - 2 * p.phi ) - p.theta ) / ( 2 * Math.PI ) ) - Math.PI / 2;
+				var id = 1 + Math.floor( ( tiles - 1 ) * ( Math.cos( Math.PI / 2 / turns ) -
```
Comment (munrocket) [inline examples/index_searchball.html]: @looeee We can trace a sphere, find spherical coordinates of a point and than take a formula in paper by Marco Abrate and Fabrizio Pollastri.

**Comment by looeee** [comment]
This is looking great, and it's way more usable than I expected. Good job. Even on my slow Moto g6 performance is great.

If this doesn't get accepted here you should still host this somewhere. 

Testing on my mobile, the images take quite a while to show up. Is that due to heavy processing, or slow download from githack? 

Would it be possible to have images appear one by one as soon as they are ready?

**Comment by munrocket** [comment]
Thanks, I did my best. But can't be better.

> the images take quite a while to show up.

Main reason is network bandwidth because it is 4k texture.
Interesting but after it downloaded to GPU all is fast. Here temp link to [github pages](https://munrocket.github.io/three.js/examples/).

> Would it be possible to have images appear one by one as soon as they are ready?

It can be a progressive loader or bunch of splited textures. I wonder if I can remap and merge it into one on the fly. Also exist a problem with `iframe`, this ball always downloaded after sidebar frame and after threejs module was downloaded and parsed. I can get this texture in sidebar and pass it through messages into iframe, seems like fun hack 😊


Here how download looks like. We can start loading before line and decrease size / split.

**Comment by Mugen87** [comment]
One of my most important conclusions from the field of information visualization is the fact that 2D UIs are more appropriate to communicate information than 3D ones in almost all cases.

3D UIs like your sphere might provide a higher degree of interaction but a classic 2D grid UI makes it much easier for users to scan through the images. The UI is less confusing and information are perceived faster (this can actually be measured). So I really think it's better to suppress the excitement for 3D in this context and go for a more common approach.

Since the result is still fun, it would be nice to share it at the forum.

**Comment by munrocket** [comment]
[The Stooges - No Fun](https://www.youtube.com/watch?v=SomQX54qhz0)

All 3d on screen actually 2d

**Comment by munrocket** [comment]
@Mugen87 

Back in the day mrdoob was more creative https://mrdoob.com/#/96/google_sphere
You don't have any searching with screenshots. Please make a searching for us.

**Comment by marcofugaro** [comment]
@Mugen87 what is your opinion about putting this directly on the homepage?

I think it's a perfect fit for the homepage exploratory focus.

**Comment by Mugen87** [comment]
@marcofugaro Sorry, nbut I think I already made my point clear in https://github.com/mrdoob/three.js/pull/18839#issuecomment-625120097.

**Comment by looeee** [comment]
>  a classic 2D grid UI makes it much easier for users to scan through the images

I love this sphere and I can see a lot of work went into making it. However, I agree with @Mugen87 here. The index page should be boring and functional. A 2D grid of images is the best possible solution. We can save the cool 3D stuff for the examples themselves.

**Comment by munrocket** [comment]
>index page should be boring and functional

Pixi.js, Babylon.js, bruno-simon, lusion.co don't think that. Not an argument.
Text based search should exist, screenshots tells nothing about what inside and as last resort.
This page already boring enough, I am not used mirrors, glow, transparency and lightning.

>We can save the cool 3D stuff for the examples themselves.

@Mugen87 will send it to showcase. Just look at this gem #18908

**Comment by Mugen87** [comment]
@munrocket Listen. I know it's frustrating when PR's are not accepted. Myself, @looeee and all other collaborators made this experience in the past. Not only with `three.js` but probably with other Open-Source projects, too. To be disappointed in the first moment is fine but it's not appropriate to react with an offensive or sarcastic undertone.

**Comment by looeee** [comment]
> I know it's frustrating when PR's are not accepted. Myself, @looeee and all other collaborators made this experience in the past. 

Yup 😅 

Those feelings of frustration have made me very aware of how important it is to treat other people's work with kindness. 

> Pixi.js, Babylon.js, bruno-simon, lusion.co don't think that.

I was curious to check these out.

[babylonjs.com/examples/](https://doc.babylonjs.com/examples/) is a simple list of thumbnails. 
[pixijs.io/examples](https://pixijs.io/examples) also a simple list, no thumbnails. 

To be honest I think the babylon.js one is the best. Maybe we can take some inspiration from that? 

2020-05-07 21_15_35-Babylon js Documentation

**Comment by munrocket** [comment]
Ok, I will stop, sorry, we just don't have alternatives.

> babylonjs is a simple list of thumbnails.

You right here, my fault, only main page weird. But gallery in Pixi is using some effects with Pixi which is kind of cool.

**Comment by munrocket** [comment]
@callumprentice we tried. Probably your turn now.
Please continue working on you PR. I am also created an [event](https://discourse.threejs.org/t/unofficial-three-js-contest-to-improve-navigation-on-sites/14992/2).

**Comment by marcofugaro** [comment]
> To be honest I think the babylon.js one is the best. Maybe we can take some inspiration from that?

@looeee Nice! I like this one also. I find it more usable than [the current proposal of the horizontal split](https://github.com/mrdoob/three.js/pull/18714#issuecomment-591676139).
The vertical split aspect ratio is better on small screens! (I have a 13'' macbook pro)

**Comment by callumprentice** [comment]
> @callumprentice we tried. Probably your turn now.

I don't mind taking this up again but I didn't have a very good experience before.

I came up with a design and it was rejected with a wholly different suggestion. I implemented that and then was told there is little chance of something like that being accepted and that I should consider ... ... ...  the design I first made. Not sure what that's about but it makes it difficult to work on if the goalposts are changing.

Plus, the code I wrote to capture screenshots with Puppeteer seems to have been rejected too.

I imagine the event you posted is helpful - once a design is solidified, it'll be less fraught to implement .

**Comment by munrocket** [comment]
I feel your pain here, that why I added reference to you in this commit.
We can turn on open-collective and hire web designer 😺

**Comment by munrocket** [comment]
Making a web page not an easy task, if someone can take responsibility for this it can be cool. Any references also good, current design pretty stylish.

**Comment by marcofugaro** [comment]
> We can turn on open-collective and hire web designer 😺

I used to call myself a web designer 🤣

**Comment by munrocket** [comment]
@marcofugaro https://github.com/mrdoob/three.js/pull/18714#issuecomment-625351168

**Comment by munrocket** [comment]
https://codepen.io/marco_fugaro :)

**Comment by callumprentice** [comment]
Would it be helpful to collectively decide on what it  should /shouldn't do? For example the absolute minimum feature set might be:
- Display page of thumbnail images from each example along with its name and link to GitHub source
- Thumbnails are generated automatically - not feasible to manually create as part of release process
- Works in a desktop browser (of course) but also works on mobile albeit with a different mechanism
- Works the same on Chrome, Firefox and Safari (Desktop/mobile versions)
- Supports filtering by example name

Some nice to haves might be:
- Filter by new/updated/existing ("what are the new examples this release?"
- Filter/search on source code too ("which examples use THREE.TrackballControls?")
- Guarantee thumbnails are representative of example (Needs additional code in each demo I suspect)
- Link directly to a single example of the page (this might be a minimum requirement)

That's a few things off the top of my head but I'm sure there are many more. Would it help to have a definitive list before we have a go at implementing something?

**Comment by munrocket** [comment]
@marcofugaro howbow dah yo

( at right we can place tags and add thin grey border, but we need generate wide screenshots )

**Comment by marcofugaro** [comment]
That's a good starting point @munrocket! I'll try to improve upon it

**Comment by munrocket** [comment]
@Mugen87 @looeee what if we will try something like this?
https://three-mesh-ui.herokuapp.com/#nested_blocks

Let's revisit our decision about 2d and also ask the repo owner.

**Comment by munrocket** [comment]
Closing in favor of #19375

**Comment by zlf2002171** [comment]
unsubscribe

2020-05-18 

zlf2002171 



发件人：munrocket <notifications@github.com>
发送时间：2020-05-18 14:24
主题：Re: [mrdoob/three.js] Make examples searchable by screenshots. (#18839)
收件人："mrdoob/three.js"<three.js@noreply.github.com>
抄送："Subscribed"<subscribed@noreply.github.com>

Closing in favor of #19375

**Comment by zlf2002171** [comment]
unsubscribe

2020-05-18 

zlf2002171 



发件人：Callum Prentice <notifications@github.com>
发送时间：2020-05-08 01:32
主题：Re: [mrdoob/three.js] Make examples searchable by screenshots. (#18839)
收件人："mrdoob/three.js"<three.js@noreply.github.com>
抄送："Subscribed"<subscribed@noreply.github.com>

Would it be helpful to collectively decide on what it should /shouldn't do? For example the absolute minimum feature set might be:
Display page of thumbnail images from each example along with its name and link to GitHub source 
Thumbnails are generated automatically - not feasible to manually create as part of release process 
Works in a desktop browser (of course) but also works on mobile albeit with a different mechanism 
Works the same on Chrome, Firefox and Safari (Desktop/mobile versions) 
Supports filtering by example name 
Some nice to haves might be:
Filter by new/updated/existing ("what are the new examples this release?" 
Filter/search on source code too ("which examples use THREE.TrackballControls?") 
Guarantee thumbnails are representative of example (Needs additional code in each demo I suspect) 
Link directly to a single example of the page (this might be a minimum requirement) 
That's a few things off the top of my head but I'm sure there are many more. Would it help to have a definitive list before we have a go at implementing something?

**Comment by munrocket** [comment]
Here final page with ball, what you lost: https://munrocket.github.io/three.js/examples/
It's loading fast without .basis and made with Three.js in Three.js style.



@mrdoob @Mugen87 you should stop this gaslighting against me and be open minded.
The thing that you ruined _tags_ in examples, _ball_ and _easy refactoring in core_ is up to you.

We found equally good solution with screenshots, but tags still just an alias that hard to maintain. Mr.doob please make tags and names in examples better, I don't have any energy to doing that.

**Comment by WestLangley** [comment]
>tags still just an alias that hard to maintain

I agree that tags will likely always be inadequate, and I would vote to remove them.

Personally, I think https://munrocket.github.io/three.js/examples/ is the best implementation, except I would prefer a full-screen quad, or scrollable quad, instead of the sphere.

>@mrdoob @Mugen87 you should stop this gaslighting against me

@munrocket Including statements like that is not a good idea. :/

**Comment by mrdoob** [comment]
Screen Shot 2020-05-25 at 2 09 31 PM

Blocked again :/

**Pull request closed without merging**